### PR TITLE
Load data using pyarrow types

### LINKED ISSUE #279: Support loading strings with the pyarrow backend to use the `string[pyarrow]` pandas dtype
Currently for string columns, pandas will load the strings as native python strings, and dask will then create a slow task to convert them all to pyarrow strings. Pandas has recently introduced support for the pyarrow string dtype, and can load strings from parquet files directly into a pandas df with the pyarrow string type by specifying `dtype_backend="pyarrow"` as an option in the `pd.read_parquet` call. 

We support passing kwargs to this function, but when generating the dask meta DataFrame from the parquet schema, we don't use pyarrow string types, and so we get a meta mismatch. So this needs to be updated here, and tested that the new dtype works with the other `from_delayed` functions for operations like crossmatching and joining where we generate the meta.

### PR DESCRIPTION
**Pull request by camposandro**
Adds the `use_pyarrow_types` argument to the `read_hipscat` interface. This allows us to pass the `dtype_backend` to the calls that read the parquet leaf files, as well as metadata, loading the data and respective schema with pyarrow types. 

Using the pyarrow backend (which is now the default!) we can load strings directly into a pandas dataframe with the pyarrow string type, avoiding the creation of slow tasks to convert python strings into pyarrow strings.

This change required me to also address #303, allowing catalogs to be created with pyarrow types in `from_dataframe`.

Closes #279 and #303.

### CODE REVIEW

**Comment by hombit** [comment]
Does it also fix #89?

**Comment by camposandro** [comment]
> Does it also fix #89?

Yes, the string conversion tasks should not be required for catalogs loaded with pyarrow types. 

I created a [notebook](https://github.com/lincc-frameworks/notebooks_lf/blob/main/sprints/2024/05_02/pyarrow-types/loading-catalogs-pyarrow-types.ipynb) that loads a chunk of ZTF sources with the "band" information, of string-type (we saw several bottlenecks happening with this column before). Using a single worker and the latest version of Dask (2024.5.0) these were the compute times I obtained:

**Previously (w/ pyarrow string conversion):** 181.35 s
**Previously (w/o pyarrow string conversion):** 88.31 s
**Currently (w/ pyarrow typed catalogs):** 83.81 s

File: src/lsdb/loaders/dataframe/from_dataframe_utils.py [hunk 1]
```diff
@@ -12,22 +13,68 @@
...
+        ddf (dd.DataFrame): A Dask DataFrame
+
+    Returns:
+        A new dask DataFrame, where columns, index and schema have been
+        converted to use pyarrow types.
+    """
+    # Convert schema types according to the backend
+    pyarrow_meta = _convert_dtypes_to_pyarrow(ddf._meta)
+    # Apply the new schema to the dask dataframe
+    ddf = ddf.astype(pyarrow_meta.dtypes)
```
Comment (smcguire-cmu) [inline src/lsdb/loaders/dataframe/from_dataframe_utils.py:51]: Can we do this conversion before making it a ddf?

File: src/lsdb/loaders/dataframe/dataframe_catalog_loader.py [hunk 2]
```diff
@@ -46,13 +47,16 @@ def __init__(
             highest_order (int): The highest partition order
             partition_size (int): The desired partition size, in number of rows
             threshold (int): The maximum number of data points per pixel
+            use_pyarrow_types (bool): If True, the data is backed by pyarrow, otherwise we keep the
+                original data types. Defaults to "pyarrow".
```
Comment (delucchi-cmu) [inline src/lsdb/loaders/dataframe/dataframe_catalog_loader.py:51]:
```suggestion
                original data types. Defaults to True.
```

File: src/lsdb/loaders/dataframe/from_dataframe.py [hunk 3]
```diff
@@ -27,6 +28,8 @@ def from_dataframe(
         threshold (int): The maximum number of data points per pixel
         margin_order (int): The order at which to generate the margin cache
         margin_threshold (float): The size of the margin cache boundary, in arcseconds
+        use_pyarrow_types (bool): If True, the data is backed by pyarrow, otherwise we keep the
+            original data types. Defaults to "pyarrow".
```
Comment (delucchi-cmu) [inline src/lsdb/loaders/dataframe/from_dataframe.py:32]:
```suggestion
            original data types. Defaults to True.
```

**Review by delucchi-cmu — APPROVED** [review]

**Review by smcguire-cmu — APPROVED** [review]

**Pull request merged**